File: BubblerApp/BubblerApp/Core/LikedPostsStore.swift
import Combine
import Foundation

/// Shared liked-post IDs so hearts stay consistent across Graph, Feed, and Search.
@MainActor
final class LikedPostsStore: ObservableObject {
    @Published private(set) var likedPostIDs = Set<String>()

    func isLiked(_ postID: String) -> Bool {
        likedPostIDs.contains(postID)
    }

    func setLiked(_ postID: String, liked: Bool) {
        if liked {
            likedPostIDs.insert(postID)
        } else {
            likedPostIDs.remove(postID)
        }
    }

    func refresh() async {
        do {
            let ids = try await APIClient.getLikedPostIDs()
            likedPostIDs = Set(ids)
        } catch {
            // Keep the local set if likes can't be loaded.
        }
    }

    func clear() {
        likedPostIDs = []
    }
}

File: bubbler_app/lib/core/storage/liked_posts_store.dart
import 'package:flutter/foundation.dart';

import '../../data/repositories/user_repository.dart';

/// Shared liked-post IDs so hearts stay consistent across Graph, Feed, and
/// Search. Mirrors Swift `LikedPostsStore`.
class LikedPostsStore extends ChangeNotifier {
  LikedPostsStore(this._userRepository);

  final UserRepository _userRepository;

  Set<String> _likedPostIds = {};

  Set<String> get likedPostIds => Set.unmodifiable(_likedPostIds);

  bool isLiked(String postId) => _likedPostIds.contains(postId);

  void setLiked(String postId, bool liked) {
    if (liked) {
      _likedPostIds = {..._likedPostIds, postId};
    } else {
      _likedPostIds = {..._likedPostIds}..remove(postId);
    }
    notifyListeners();
  }

  /// Reloads liked IDs from the server. Keeps the local set on failure.
  Future<void> refresh() async {
    try {
      final ids = await _userRepository.getLikedPostIds();
      _likedPostIds = ids.toSet();
      notifyListeners();
    } on Object {
      // Keep the local set if likes can't be loaded.
    }
  }

  void clear() {
    _likedPostIds = {};
    notifyListeners();
  }
}
